agent/lm_agent/tokenstat.py

```diff
@@ -1,219 +1,223 @@
-    used: int
-    total: int
-    used_licenses: typing.List
+    @abc.abstractclassmethod
+    def get_report_item(self, product_feature: str):
+        """Parse license server output into a report item for the indicated feature."""
+        raise NotImplementedError
 
-    @classmethod
-    def from_stdout(cls, product, parse_fn, tool_name, stdout):
-        """
-        Create a LicenseReportItem by parsing the stdout from the program that
-        produced it.
-        """
-        parsed = parse_fn(stdout)
-        return cls(
-            tool_name=tool_name,
-            product_feature=f"{product}.{parsed['total']['feature']}",
-            used=parsed["total"]["used"],
-            total=parsed["total"]["total"],
-            used_licenses=parsed["uses"],
+
+class FlexLMLicenseServer(LicenseServerInterface):
+    """Extract license information from FlexLM license server."""
+
+    def __init__(self, license_servers: typing.List[str]):
+        self.license_servers = license_servers
+        self.parser = flexlm.parse
+
+    def get_commands_list(self):
+        """Generate a list of commands with the available license server hosts."""
+
+        host_ports = [(server.split(":")[1], server.split(":")[2]) for server in self.license_servers]
+        commands_to_run = []
+        for host, port in host_ports:
+            command_line = f"{settings.LMUTIL_PATH} lmstat -c {port}@{host} -f"
+            commands_to_run.append(command_line)
+        return commands_to_run
+
+    async def get_output_from_server(self, product_feature: str):
+        """Override abstract method to get output from FlexLM license server."""
+
+        # get the list of commands for each license server host
+        commands_to_run = self.get_commands_list()
+
+        # run each command in the list, one at a time, until one succeds
+        for cmd in commands_to_run:
+            feature = product_feature.split(".")[1]
+            cmd = cmd + f" {feature}"
+            proc = await asyncio.create_subprocess_shell(
+                cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.STDOUT
+            )
+
+            # block until a check at this host:port succeeds or fails
+            stdout, _ = await asyncio.wait_for(proc.communicate(), TOOL_TIMEOUT)
+            output = str(stdout, encoding=ENCODING)
+
+            if proc.returncode != 0:
+                logger.error(f"Error: {output} | Return Code: {proc.returncode}")
+                raise RuntimeError("None of the checks for FlexLM succeeded")
+            return output
+
+    async def get_report_item(self, product_feature: str):
+        """Override abstract method to parse FlexLM license server output into License Report Item."""
+
+        server_output = await self.get_output_from_server(product_feature)
+        parsed_output = self.parser(server_output)
+
+        # raise exception if parser didn't output license information
+        if not parsed_output["total"]:
```

Comment:
Should we use `get()` here instead or can we trust that the parser will always return output with a "total" item?

```suggestion
        if not parsed_output.get("total"):
```

Reply:
Should we also check for "uses" and make sure that the "total" sub-dict includes its own "used" and "total"?

Reply:
The `flexlm` parser will always return the `total` and `uses`.
If nothing was matched in the output, the `total` will be `None` and the `uses` an empty list.
If the total is not None, then it will always have the `total` and `used` inside.

But I'll add the validations anyway, better play it safe.